broadcaster/_backends/pulsar.py

```diff
@@ -1,77 +1,119 @@
 import anyio
+import asyncio
 import logging
 import typing
 from urllib.parse import urlparse
 import pulsar
 from broadcaster._base import Event
 from .base import BroadcastBackend
 
+logger = logging.getLogger(__name__)
 
 class PulsarBackend(BroadcastBackend):
-    def __init__(self, url: str):
+    def __init__(self, url: str, max_queue_size: int = 1000):
         parsed_url = urlparse(url)
         self._host = parsed_url.hostname or "localhost"
         self._port = parsed_url.port or 6650
         self._service_url = f"pulsar://{self._host}:{self._port}"
         self._client = None
-        self._producer = None
-        self._consumer = None
+        self._producers = {}
+        self._consumers = {}
+        self._receiver_tasks = {}
+        self._shared_queue = asyncio.Queue(maxsize=max_queue_size)
 
     async def connect(self) -> None:
         try:
-            logging.info("Connecting to brokers")
+            logger.info("Connecting to Pulsar brokers")
             self._client = await anyio.to_thread.run_sync(
                 lambda: pulsar.Client(self._service_url)
             )
-            self._producer = await anyio.to_thread.run_sync(
-                lambda: self._client.create_producer("broadcast")
-            )
-            self._consumer = await anyio.to_thread.run_sync(
-                lambda: self._client.subscribe(
-                    "broadcast",
-                    subscription_name="broadcast_subscription",
-                    consumer_type=pulsar.ConsumerType.Shared,
-                )
-            )
-            logging.info("Successfully connected to brokers")
+            logger.info("Successfully connected to Pulsar brokers")
         except Exception as e:
-            logging.error(e)
+            logger.error(f"Error connecting to Pulsar: {e}")
             raise e
 
     async def disconnect(self) -> None:
-        if self._producer:
-            await anyio.to_thread.run_sync(self._producer.close)
-        if self._consumer:
-            await anyio.to_thread.run_sync(self._consumer.close)
+        # Cancel all receiver tasks
+        for task in self._receiver_tasks.values():
+            task.cancel()
+        
+        # Wait for all receiver tasks to complete
+        await asyncio.gather(*self._receiver_tasks.values(), return_exceptions=True)
+        
+        # Prepare coroutines for closing producers and consumers
+        close_coros = [
+            anyio.to_thread.run_sync(producer.close)
+            for producer in self._producers.values()
+        ] + [
+            anyio.to_thread.run_sync(consumer.close)
+            for consumer in self._consumers.values()
+        ]
+        
+        # Add client close coroutine if client exists
         if self._client:
-            await anyio.to_thread.run_sync(self._client.close)
+            close_coros.append(anyio.to_thread.run_sync(self._client.close))
```

Comment:
The order probably do matter here. The `_client.close()` should be after the producers and consumers are closed